sw/device/silicon_creator/lib/manifest.h

```diff
@@ -0,0 +1,210 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef OPENTITAN_SW_DEVICE_SILICON_CREATOR_LIB_MANIFEST_H_
+#define OPENTITAN_SW_DEVICE_SILICON_CREATOR_LIB_MANIFEST_H_
+
+#include <assert.h>
+#include <stddef.h>
+
+#include "sw/device/silicon_creator/lib/error.h"
+// FIXME: Move sigverify to sw/device/silicon_creator/lib
+#include "sw/device/silicon_creator/mask_rom/sig_verify_keys.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif  // __cplusplus
+
+/**
+ * Manifest for boot stage images stored in flash.
+ *
+ * OpenTitan secure boot, at a minimum, consists of four boot stages: ROM,
+ * ROM_EXT, BL0, and Kernel. ROM is stored in the read-only Mask ROM while
+ * remaning stages are stored in flash. This structure must be placed at the
+ * beginning of ROM_EXT and BL0 images so that ROM and ROM_EXT can verify
+ * integrity and authenticity of the next stage before handing over execution.
+ *
+ * Use of this struct for stages following BL0 is optional.
+ */
+typedef struct manifest {
+  /**
+   * Manifest identifier.
+   */
+  uint32_t identifier;
+  /**
+   * FIXME: remove this field.
+   */
+  uint32_t reserved0;
+  /**
+   * Image signature.
+   *
+   * The signed region of an image starts at `image_length` and ends at the
+   * end of the image. The start and the length of the signed region can be
+   * obtained using `manifest_signed_region_get`.
+   */
+  sigverify_rsa_buffer_t signature;
+  /**
+   * Image length in bytes.
+   */
+  uint32_t image_length;
+  /**
+   * FIXME: Replace with max_version, min_version.
+   */
+  uint32_t image_version;
+  /**
+   * Image timestamp.
+   */
+  uint64_t image_timestamp;
+  /**
+   * Exponent of the signer's RSA public key.
+   */
+  uint32_t exponent;
+  /**
+   * FIXME: remove this field.
+   */
+  uint32_t reserved1;
+  /**
+   * FIXME: Replace these with binding_tag and max_key_version.
+   */
+  uint32_t usage_constraints[8];
+  uint32_t lockdown_info[4];
+  /**
+   * Modulus of the signer's RSA public key.
+   */
+  sigverify_rsa_buffer_t modulus;
+  /**
+   * Extension fields.
+   * FIXME: Remove these until we have a clear use-case.
+   */
+  uint32_t extension0_offset;
+  uint32_t extension0_checksum;
+  uint32_t extension1_offset;
+  uint32_t extension1_checksum;
+  uint32_t extension2_offset;
+  uint32_t extension2_checksum;
+  uint32_t extension3_offset;
+  uint32_t extension3_checksum;
+} manifest_t;
+
+static_assert(offsetof(manifest_t, identifier) == 0, "");
```

Comment:
Although the intention is completely understandable from the code, what kind of error does the compiler emit should one of these be wrong?  (and: should we then fill in an error string in the assert?)

Reply:
Oh, thanks for catching this. I forgot to update messages but now I'm not sure if we should have a message and if we should if they should be identical or unique.
Without a message:
```
In file included from ../sw/device/silicon_creator/rom_exts/rom_ext.c:12:
/workspace/worktrees/opentitan/rom-ext-manifest/sw/device/silicon_creator/lib/manifest.h:90:1: error: static_assert failed due to requirement '__builtin_offsetof(struct manifest, identifier) == 4' ""
static_assert(offsetof(manifest_t, identifier) == 4, "");
^             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.../sw/device/lib/base/freestanding/assert.h:23:23: note: expanded from macro 'static_assert'
#define static_assert _Static_assert
                      ^
1 error generated.
```
With message = "foo":
```
In file included from ../sw/device/silicon_creator/rom_exts/rom_ext.c:12:
/workspace/worktrees/opentitan/rom-ext-manifest/sw/device/silicon_creator/lib/manifest.h:90:1: error: static_assert failed due to requirement '__builtin_offsetof(struct manifest, identifier) == 4' "foo"
static_assert(offsetof(manifest_t, identifier) == 4, "foo");
^             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.../sw/device/lib/base/freestanding/assert.h:23:23: note: expanded from macro 'static_assert'
#define static_assert _Static_assert
                      ^
1 error generated.```

Reply:
I think the message should be something like "Unexpected offset for ${struct_member}".

Reply:
Maybe this is sufficiently common that we want a macro like

```c
#define ASSERT_OFFSET(type, member, offset) \
  static_assert(offsetof(type, member) == UINT32_C(offset), "Unexpected offset for " #type "." #member)
```

Reply:
Added `OT_ASSERT_MEMBER_OFFSET` and `OT_ASSERT_SIZE` to `lib/base/macros.h`.